### ISSUE TITLE
Bug: sometimes a window on the focused monitor is focused when a window of the same app on a different monitor was requested to focus

### ISSUE THREAD

**Issue by Jasperswaagman** [nikitabobko/AeroSpace#101]
Hey Nikita, I'm experiencing the following behavior and I've no idea if it's intended or not. Whenever I've two instances of an application open on different monitors/workspaces. It will focus the instance of the app I last used. I'll try to explain it with an example:

I've a three monitor setup: `1 | 2 | *0`

* Workspace 1 contains a kitty window.
* Workspace 2 contains a kitty window
* Workspace 0 contains something else, say firefox.

I've the following keybinds to go the specific workspaces:
```
alt-1 = 'workspace 1'
alt-2 = 'workspace 2'
```

I want to go to workspace 1 so I press `alt-1`. I end up at workspace 2. If I press `alt-1` again I'll go to workspace 1 as expected. 

Is this working as intended? I would like to be able to actually focus a specific window/instance of an app.

Love the app so far, is there a way we can buy you a coffee or beer?

**Comment by johnallen3d** [comment]
I've experienced a very similar issue with Arc browser. I've had similar issues with Arc and Yabai in the past so I thought it was an Arc specific problem. Perhaps not.

**Comment by nikitabobko** [comment]
AFAIK it's a bug in Apple accessibility API. If not mistaken, I saw the special handling of cases like this in yabai and/or Amethyst. But right now I can't find it in their codebases and issue trackers (if somebody finds it, it'd be helpful to post link here)

I suspect that this bug might go away if you disable "Displays have separate spaces", but I didn't test it

**Comment by Jasperswaagman** [comment]
> I suspect that this bug might go away if you disable "Displays have separate spaces", but I didn't test it
That does the trick, awesome! Thank you!

**Comment by nikitabobko** [comment]
Should be fixed in 0.11.1-Beta

It'd would be cool if somebody can confirm

**Comment by alexnguyennn** [comment]
I just gave this a try to confirm on `0.11.2-Beta`: (with all other automation services e.g. `sketchybar`, `hammerspoon` disabled and `Displays have separate spaces` enabled) it seems to me like I'm still running into the above described issue.

I run 3 external monitors and swapping workspaces seems to still have a different instance of the app on another monitor steal focus specifically when src/destination workspace contain different active apps. Can't seem to see a pattern otherwise (for me it isn't always the last selected, just another visible instance) When changing workspaces where src/destination have the same active app this issue doesn't seem to occur.

Silver lining is that the previous workaround of turning `Displays have separate spaces` back off still works, though it would be a nice-to-have to be able to turn it off so that full screen doesn't black all other monitors out 😅

**Comment by nikitabobko** [comment]
@alexnguyennn thanks for checking! I confirm that the problem is still reproducible with 3 monitors. 

I normally use 1-2 monitors, and the problem appears to be fixed only for 2 monitors

It's unbelievable how buggy and problematic the API is

**Comment by alimg** [comment]
I encountered this problem on `AeroSpace v0.11.2-Beta 5559f963f6fed4d7b2deff8e064350e18e184c78` today using just 1 external monitor.

I am switching from Amethyst after using for years and problem feels very familiar. But I think it is largely solved nowadays, thanks to this commit:
https://github.com/ianyh/Amethyst/pull/1530

**Comment by jmml97** [comment]
This is happening to me. My setup is as follows:

- Monitor 1: space 1 with Visual Studio Code, space 2 with Arc Browser
- Monitor 1: space 3 with Visual Studio Code, space 4 with Arc Browser

If Monitor 1 is on space 1 and Monitor 2 on space 3, when I change to space 4, Monitor 1 also changes to space 2.

The problem seems to happen when switching to a space on the monitor different to the one currently focused.

My Aerospace version:

```cli
❯ aerospace --version
aerospace CLI client version: 0.14.2-Beta 0cb8dbdfc5ee73b8cbc200f175f467ebead55201
AeroSpace.app server version: 0.14.2-Beta 0cb8dbdfc5ee73b8cbc200f175f467ebead55201
```

**Comment by johnallen3d** [comment]
PSA - this "bug" popped back up for me after upgrading to Sequoia. It turns out the upgrade re-enabled the "Displays have separate spaces" settings. 🙄

**Comment by docmg** [comment]
I’m experiencing the same bug using multiple Chrome windows. Turning off “Displays have separate spaces” helps, but only after disabling Aerospace and only then restarting the system.

The first time I tested this, I didn’t disable Aerospace before restarting. As a result, the second monitor wasn’t properly recognized by Aerospace, and any windows I moved to it just disappeared. :)

**Comment by pourmand1376** [comment]
I can confirm that this bug is fixed after disabling separate spaces.

But finding this issue was a little hard. Is there any way to make it more visible for new users? (Like to add it in documentation or pin it in the issues).

**Comment by ok-nick** [comment]
I've compiled some information for how amethyst/yabai/hammerspoon solves this issue:
[amethyst code](https://github.com/ianyh/Amethyst/pull/1530/files) (from @alimg)
[yabai code](https://github.com/koekeishiya/yabai/blob/f4ada8d557aeb9c8ba0d031ee6f645acffff50fd/src/window_manager.c#L1281-L1323)
https://github.com/Hammerspoon/hammerspoon/issues/370 (probably the most informative)

It looks like the creator of yabai (and later amethyst) solved it by using private APIs described [here](https://github.com/Hammerspoon/hammerspoon/issues/370#issuecomment-545545468). However, another user mentions they were able to activate the application, then focus the window after 0.001 seconds, and have this functionality work reliably, [found here](https://github.com/Hammerspoon/hammerspoon/issues/370#issuecomment-615535897).

I think it's worth adding the delay for now so that the "displays have separate spaces" setting doesn't need to be disabled, but overall it seems like the private API will give the most robust functionality. Perhaps aerospace should introduce settings to toggle the use of private APIs such as this?

---

I tried testing the non-private API theory by activating the app, delaying for 1s, then raising the window, but it seems hammerspoon does a bit more [magic under the hood](https://github.com/Hammerspoon/hammerspoon/blob/931d38acbd7988dc2ec87a9c095bc50bd83ca584/extensions/window/window.lua#L507-L537) to focus a window:

```lua
function objectMT.focus(self)
  local app=self:application()
  if app then
    self:becomeMain()
    app:_bringtofront()
    if app:bundleID()=='com.apple.finder' then --workaround for the desktop taking over
      -- it may look like this should ideally go inside :becomeMain(), but the problem is actually
      -- triggered by :_bringtofront(), so the workaround belongs here
      if desktopFocusWorkaroundTimer then desktopFocusWorkaroundTimer:stop() end
      desktopFocusWorkaroundTimer=timer.doAfter(0.3,function()
        -- 0.3s comes from https://github.com/Hammerspoon/hammerspoon/issues/581
        -- it'd be slightly less ugly to use a "space change completed" callback (as per issue above) rather than
        -- a crude timer, although that route is a lot more complicated
        self:becomeMain()
        desktopFocusWorkaroundTimer=nil --cleanup the timer
      end)
      self:becomeMain() --ensure space change actually takes place when necessary
    end
  end
  return self
end
```

In the code, `self:becomeMain()` activates the app and sets the `NSAccessibilityMainAttribute` attribute, [here](https://github.com/Hammerspoon/hammerspoon/blob/931d38acbd7988dc2ec87a9c095bc50bd83ca584/Hammerspoon/HSuicore.m#L848-L850) (I’m not sure why they use `NSAccessibilityMainAttribute` instead of `kAXMainAttribute`, maybe it’s the same thing?). Next, they call `app:_bringtofront()` which further calls `SetFrontProcessWithOptions`, [here](https://github.com/Hammerspoon/hammerspoon/blob/931d38acbd7988dc2ec87a9c095bc50bd83ca584/Hammerspoon/HSuicore.m#L257-L264). Then finally, they do a bit more [crazy work to finish everything off](https://github.com/Hammerspoon/hammerspoon/blob/931d38acbd7988dc2ec87a9c095bc50bd83ca584/extensions/window/window.lua#L522-L534). The 0.3s delay looks a bit concerning for a window manager. I’m not quite sure how much of this code is necessary–I’d assume all of it to be robust–but it’s quite magical. 

@nikitabobko any opinions on this?

**Comment by nikitabobko** [comment]
@ok-nick thanks for the pointers.

**Below are my thoughts out loud. I don't know the final decision yet.**

This issue is irritating. At least it was for me, until I discovered the workaround to disable "Displays have separate Spaces", and I personally is fine to disable it, since Spaces are cursed in some many different ways.

But I think I'm _maybe_ fine to use private APIs specifically for this issue, but only when:
- "Displays have separate Spaces" is enabled. It's possible to check it with `NSScreen.screensHaveSeparateSpaces`
- The window that we want to focus is located on a different monitor
- _Maybe_ add an option in config to allow usage of private API in this case, but I don't like to overwhelm the config with such options. I want everything to work right out of the box. Agghhh
- Assignment for the readers: think about other ways to restrict the usage of private API?

I like the "tagline" in the README that `_AXUIElementGetWindow` is the only private API that we use right now, and if we use another private API, it won't longer be true, and we'd have to change the tagline...

The other concern is maintability. We should answer the question "what will we do, when this API stops working?". The answer is probably: "we will _just_ reintroduce this #101 bug", and fallback to only using the public API as we did before.

IDK. Other thoughts are welcome

**Comment by nikitabobko** [comment]
I've come up with the following idea:

1. Before using the private API, AeroSpace guards itself by setting a [UserDefaults](https://developer.apple.com/documentation/foundation/userdefaults) `i-am-about-to-use-private-api-to-reliably-focus-window-on-another-monitor` boolean flag to `true`
2. AeroSpace uses the API
3. If we survived the call, AeroSpace resets `i-am-about-to-use-private-api-to-reliably-focus-window-on-another-monitor` back to `false`

On AeroSpace startup (in `initAppBundle` function), AeroSpace checks if `i-am-about-to-use-private-api-to-reliably-focus-window-on-another-monitor` set to `true`. If it's set to `true`, it means that we probably crashed by using the private API. In that case, AeroSpace sets another UserDefaults boolean flag `i-will-never-use-this-focus-window-on-another-monitor-broken-private-api-from-now-on` to `true` and doesn't use the private API anymore.

Open question: should AeroSpace show the dialog and explain to user what happened when `i-will-never-use-this-focus-window-on-another-monitor-broken-private-api-from-now-on` is set to `true`? Probably yes. With the prompt to report a GitHub issue and mention macOS version.

And as suggested before, the API should be used only when:
- "Displays have separate Spaces" is enabled
- The window that we want to focus is located on a different monitor

PRs that implement the idea are welcome. The issue is not the top-priority for me at this moment.

**Comment by ok-nick** [comment]
Hm, I like the idea of adding fallbacks, but there's also other factors to consider. What if the computer crashes, but not due to the private API? The flag would be set and Aerospace would default to the old behavior, which can be confusing. Another potential case is if this API were to change, would it actually crash the program? Report an error? Cause different behavior? Or just silently fail?

For those reasons, I think Aerospace should just supply a (heavily disclaimed) config option to toggle this behavior. At least if you are opt-in, you know the first place to check when there's a problem. Perhaps there can be a "safe mode" option in the menu bar to disable "unstable" features like this.

It's unfortunate because you want to provide a robust and stable experience by default, but then there's things like this. Ideally this feature would be enabled by default, but that's the tradeoff.

**Comment by nikitabobko** [comment]
Other suggestions:
1. When the private API crash detection detects a crash, instead of not using the private API from now on, ask the user about it in the dialog
2. When the user tries to focus the same monitor 3 times in a row, show a prompt not to use the private API anymore (guard from silent failures)

**Comment by Zurki** [comment]
Bumping this. It makes me furious, as I often have one private and one business space on Chrome. It sometimes bugs out and switches in an infinite loop.

**Comment by gnohj** [comment]
Luckily, the only app I run multiple instances of across monitors is the browser—so I’ve got Zen on one and Chrome on the other. Good enough for now 🙃

**Comment by AnirudhG07** [comment]
I am not sure if my issue is exactly this, but def similar(I have disabled "Displayed have separate spaces".
I have Terminal open in workspace 1 and 2. and browser in Workspace 2 as well.

When i press `alt-1` to go to terminal, If I am currently focused on terminal in `2`, it goes fine. But if i am focused on the browser in `2`, then the focus shifts to terminal in `1` but the window shows up of browser.

How do i fix this? I enabled and disabled the above seperate spaces option but this issue still persists and is VERY annoying. If my browser is at `3`, then totally random things happen, window goes 3 ->2 or 3->1 when i press `alt-1` and focus somewhere else. Its just become unusable at this point. Please help

**Comment by cleanshorts68-tech** [comment]
> I am not sure if my issue is exactly this, but def similar(I have disabled "Displayed have separate spaces". I have Terminal open in workspace 1 and 2. and browser in Workspace 2 as well.
> 
> When i press `alt-1` to go to terminal, If I am currently focused on terminal in `2`, it goes fine. But if i am focused on the browser in `2`, then the focus shifts to terminal in `1` but the window shows up of browser.
> 
> How do i fix this? I enabled and disabled the above seperate spaces option but this issue still persists and is VERY annoying. If my browser is at `3`, then totally random things happen, window goes 3 ->2 or 3->1 when i press `alt-1` and focus somewhere else. Its just become unusable at this point. Please help

Have you managed to figure out this problem somehow?

**Comment by AnirudhG07** [comment]
> > I am not sure if my issue is exactly this, but def similar(I have disabled "Displayed have separate spaces". I have Terminal open in workspace 1 and 2. and browser in Workspace 2 as well.
> > When i press `alt-1` to go to terminal, If I am currently focused on terminal in `2`, it goes fine. But if i am focused on the browser in `2`, then the focus shifts to terminal in `1` but the window shows up of browser.
> > How do i fix this? I enabled and disabled the above seperate spaces option but this issue still persists and is VERY annoying. If my browser is at `3`, then totally random things happen, window goes 3 ->2 or 3->1 when i press `alt-1` and focus somewhere else. Its just become unusable at this point. Please help
> 
> Have you managed to figure out this problem somehow?

Oh no no, i just wish it doesnt come. if it comes, i log out and login, and then it works for sometime(like days) till again its screwed and again i logout.
not a solution but a temporary not ideal workaround to live with it i guess

**Comment by brandonchinn178** [comment]
+1 to fixing this as well. I can't turn off "Displays have separate Spaces" because https://github.com/nikitabobko/AeroSpace/issues/594 bothers me more than this

**Comment by rwblokzijl** [comment]
I have 2 other issues that i think might be related:

1. an issue where opening a new browser (brave) window will sometimes cause the window to open on the wrong workspace (where another browser lives) instead of the one i just selected.
    a. Can be avoided by selecting the workspace multiple times before opening the brower.
    b. This is especially bad if i have another browser window in native fullscreen. This causes the window to open at the fullscreen window,  overlays the new window, then hides it when you select the workspace.
		- Can be avoided by first selecting a 3rd browser window, then selecting the target workspace (multiple times), and only then opening the new browser window.
2. An issue where closing a browser window will focus another browser window on another workspace instead of staying on the current (now empty) workspace.

**Comment by yohanduartep** [comment]
> I have 2 other issues that i think might be related:
> 
>     1. an issue where opening a new browser (brave) window will sometimes cause the window to open on the wrong workspace (where another browser lives) instead of the one i just selected.
>        a. Can be avoided by selecting the workspace multiple times before opening the brower.
>        b. This is especially bad if i have another browser window in native fullscreen. This causes the window to open at the fullscreen window,  overlays the new window, then hides it when you select the workspace.
>        - Can be avoided by first selecting a 3rd browser window, then selecting the target workspace (multiple times), and only then opening the new browser window.
> 
>     2. An issue where closing a browser window will focus another browser window on another workspace instead of staying on the current (now empty) workspace.

I also have these issues.
Also, when i have 2 different tabs of the same browser (1 monitor 1 and 1 monitor 2), if i focus e.g. terminal on monitor 1 and e.g. files on monitor 2, when i focus workspace where one of the browser is, it focus both browsers workspace.

Better explanation of example
firefox workspace 1 (monitor 1)
terminal workspace 2 (monitor 1) <- focused
firefox workspace 3 (monitor 2) 
files workspace 4 (monitor 2) <- focused

When i try to switch to workspace 1 or 3 (only one of it), i get
firefox workspace 1 (monitor 1) <- focused
terminal workspace 2 (monitor 1) 
firefox workspace 3 (monitor 2) <- focused
files workspace 4 (monitor 2)

**Comment by santip** [comment]
Hi @nikitabobko, I too am affected by this, currently using aerospace in a 3 display setup, but needing Display have separate spaces enabled for two main reasons:
1. Support for multi-tasking while leveraging native fullscreen apps
2. Proper access to the menubar (I auto-hide the menubar and use sketchybar, with the separate spaces option disabled it's imposible to interact with any menu of apps living in displays other than the main monitor.

Was the private API alternate path ever implemented? is there a setting or even a patch I could apply to rebuild aerospace with this option enabled? Thanks a lot for a great window manager, by far the most stable and reliable I've tried (and I've tried quite a few), so I understand and appreciate the benefits of the philosophy to rely only on public APIs, but in this particular instance it's a broken feature, and it happens to be a pretty fundamental feature (focusing windows).

Thanks again!

> I've come up with the following idea:
> 
> 1. Before using the private API, AeroSpace guards itself by setting a [UserDefaults](https://developer.apple.com/documentation/foundation/userdefaults) `i-am-about-to-use-private-api-to-reliably-focus-window-on-another-monitor` boolean flag to `true`
> 2. AeroSpace uses the API
> 3. If we survived the call, AeroSpace resets `i-am-about-to-use-private-api-to-reliably-focus-window-on-another-monitor` back to `false`
> 
> On AeroSpace startup (in `initAppBundle` function), AeroSpace checks if `i-am-about-to-use-private-api-to-reliably-focus-window-on-another-monitor` set to `true`. If it's set to `true`, it means that we probably crashed by using the private API. In that case, AeroSpace sets another UserDefaults boolean flag `i-will-never-use-this-focus-window-on-another-monitor-broken-private-api-from-now-on` to `true` and doesn't use the private API anymore.
> 
> Open question: should AeroSpace show the dialog and explain to user what happened when `i-will-never-use-this-focus-window-on-another-monitor-broken-private-api-from-now-on` is set to `true`? Probably yes. With the prompt to report a GitHub issue and mention macOS version.
> 
> And as suggested before, the API should be used only when:
> 
> * "Displays have separate Spaces" is enabled
> * The window that we want to focus is located on a different monitor
> 
> PRs that implement the idea are welcome. The issue is not the top-priority for me at this moment.

**Comment by jimmyff** [comment]
I think I've narrowed the/a trigger for this: **it happens whenever the same app has windows on more than one monitor.** Asking AeroSpace to focus a specific window of that app - even by exact `--window-id` intermittently focuses the app's *other* window on the *other* monitor instead.

### Environment
- AeroSpace `0.20.3-Beta` (`6dde91b`)
- macOS Darwin 25.5.0
- 2 monitors: `Built-in Retina Display` (id 1) + `LG HDR WQHD+` (id 2)
- Workspace→monitor force-assignment is set, but **not required to repro** — the only thing that matters is one app with a window on each monitor.

### Minimal repro
1. Open **two windows of one app**, one on each monitor (any app -I hit it with Chromium).
2. Find their window-ids:
   ```
     aerospace list-windows --all --format '%{window-id} | %{app-name} | mon=%{monitor-id}'
   ```
3. Focus each by its exact id and print what actually got focused:
   ```bash
   A=<window-id on monitor 1>
   B=<window-id on monitor 2>
   for id in "$A" "$B"; do
     aerospace focus --window-id "$id"
     echo "asked $id -> got $(aerospace list-windows --focused --format '%{window-id} (%{app-name}, mon %{monitor-id})')"
   done
   ```

**Expected:** each line echoes back the id I asked for.

**Actual (sample):**
```
asked 93  -> got 733 (Chromium, mon 1)    # asked for the LG window, got the built-in one
asked 733 -> got 93  (Chromium, mon 2)    # and vice-versa
```

So `focus --window-id <X>` focuses a *different* window of the same app. The knock-on effect is what most people actually notice: `workspace <name>`, `focus left/right`, and `focus-monitor` toward that monitor all get yanked to the wrong monitor too (they ultimately focus that window), which presents as *"switching monitors changed the other monitor's workspace."*

### Notes that may help
- Workspaces whose visible window is a **unique** app focus correctly every time; only the **duplicated-app** workspace misfires - strong signal it's same-app window resolution, not navigation logic.
- **Recency-dependent**: whichever of the two windows was focused most recently tends to win, which is why it feels random/intermittent.
- Fires even when the app's other window is on a **hidden** workspace (not currently visible).
- **Workaround that fully fixes it for me:** keep all of an app's windows on a single monitor (no cross-monitor split). The moment I moved the second Chromium window onto the same monitor as the first, every monitor hop worked.

Claudes' guess is that `AXRaise` on the specific `AXWindow` is being overridden by macOS raising the app's "key"/frontmost window on activation (Chromium/Electron seem especially prone).

**Comment by hunner** [comment]
Claude and I spent a couple of days on this with a two-monitor setup (macos 26.6, Apple Silicon, Firefox with a window on each monitor for reproducing) and can confirm most of @jimmyff's findings as well. We could not fix it, but instrumented it enough to see what is happening better, and wanted to put notes here in case it saves someone else the time.

The mechanism `nativeFocus` ultimately calls `nsApp.activate()`. That is asynchronous, and as activation completes macos restores the app's own focused window. Anything AeroSpace sets beforehand is undone. With one app holding windows on two monitors, the app's choice wins and focus may land on the monitor I did not have focused.

Things that do not fix it, all measured by logging what macos had focused against what was requested, per command:

- Raising before activating, which is what the code does today
- Raising after activating as well, on both sides
- Setting `kAXFocused` rather than only `kAXMain`, which matters because `kAXMain` marks the main window and raising only affects z-order, but it still is not enough
- Asserting the window from `NSWorkspace.didActivateApplicationNotification`, on the theory that acting after the notification would be acting last. It is not.

So macos settles the key window after every hook we can observe. I don't know if this is fixable through public AX, which presumably is why yabai has SkyLight.

One part of this that is probably worth fixing anyway... `updateFocusCache` adopts whatever macos focused, and `setFocus` calls `setActiveWorkspace`, so when the app's restored window happens to live on a workspace that is not visible, that workspace gets pulled onto its monitor. Focus then settles back where it belongs and the workspace change stays behind. That is, when I was focused on one monitor and switch to a space with firefox, the other monitor will sometimes switch to a space with firefox also. Skipping adoption for windows parked off-screen fixes the workspace relocation, even though it does not fix the focus problem.

Also kind of related, we found that the fast path in `nativeFocus` is gated on `!NSScreen.screensHaveSeparateSpaces || monitorInfos.count == 1`, which assumes this only bites with "Displays have separate Spaces" enabled. We hit it with that off too, and the guide recommends turning it off, so the shortcut applies to people following the docs.

I'll continue hunting. Hoping apple exposes some more hooks that we can use eventually!
